frame/nft-fractionalisation/src/lib.rs

```diff
@@ -0,0 +1,294 @@
+
+#[frame_support::pallet]
+pub mod pallet {
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	use frame_support::{
+		dispatch::DispatchResult,
+		sp_runtime::traits::{
+			AccountIdConversion, AtLeast32BitUnsigned, IntegerSquareRoot, SaturatedConversion,
+			Saturating, StaticLookup, Zero,
+		},
+		traits::{
+			fungibles::{
+				metadata::Mutate as MutateMetadata, Create, Inspect, InspectEnumerable, Mutate,
+				Transfer,
+			},
+			tokens::nonfungibles_v2::{
+				Inspect as NonFungiblesInspect, Transfer as NonFungiblesTransfer,
+			},
+			Currency,
+		},
+		PalletId,
+	};
+
+	pub use pallet_nfts::Incrementable;
+
+	pub type BalanceOf<T> =
+		<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+	pub type AssetIdOf<T> =
+		<<T as Config>::Assets as Inspect<<T as frame_system::Config>::AccountId>>::AssetId;
+	pub type AssetBalanceOf<T> =
+		<<T as Config>::Assets as Inspect<<T as frame_system::Config>::AccountId>>::Balance;
+
+	pub type AccountIdLookupOf<T> = <<T as frame_system::Config>::Lookup as StaticLookup>::Source;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		type Currency: Currency<Self::AccountId>;
+
+		/// Identifier for the collection of item.
+		type CollectionId: Member + Parameter + MaxEncodedLen + Copy + Incrementable;
+
+		/// The type used to identify a unique item within a collection.
+		type ItemId: Member + Parameter + MaxEncodedLen + Copy;
+
+		type AssetBalance: AtLeast32BitUnsigned
+			+ codec::FullCodec
+			+ Copy
+			+ MaybeSerializeDeserialize
+			+ sp_std::fmt::Debug
+			+ Default
+			+ From<u64>
+			+ IntegerSquareRoot
+			+ Zero
+			+ TypeInfo
+			+ MaxEncodedLen;
+
+		type AssetId: Member
+			+ Parameter
+			+ Default
+			+ Copy
+			+ codec::HasCompact
+			+ From<u32>
+			+ MaybeSerializeDeserialize
+			+ MaxEncodedLen
+			+ PartialOrd
+			+ TypeInfo;
+
+		type Assets: Inspect<Self::AccountId, AssetId = Self::AssetId, Balance = Self::AssetBalance>
+			+ Create<Self::AccountId>
+			+ InspectEnumerable<Self::AccountId>
+			+ Mutate<Self::AccountId>
+			+ MutateMetadata<Self::AccountId>
+			+ Transfer<Self::AccountId>;
+
+		type Items: NonFungiblesInspect<
+				Self::AccountId,
+				ItemId = Self::ItemId,
+				CollectionId = Self::CollectionId,
+			> + NonFungiblesTransfer<Self::AccountId>;
+
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		#[pallet::constant]
+		type BuybackThreshold: Get<u32>;
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn assets_minted)]
+	// TODO: query amount minted from pallet assets instead of storing it locally.
+	// Add a public getter function to pallet assets.
+	pub type AssetsMinted<T: Config> =
+		StorageMap<_, Twox64Concat, AssetIdOf<T>, AssetBalanceOf<T>, OptionQuery>;
```

Comment:
Is there a specific reason why is using the `Twox64Concat` hashing algorithm here? I wonder if it's not more secure to use Blake2

Reply:
Do we need a cryptographic hash function here? This is just a reference from NFT to `AssetId`